tweaks (#12)
fsprojects/fsharp-cheatsheet: docs/fsharp-cheatsheet.md

    let s, f, d, bi = 4.14F, 4.14, 0.7833M, 9999I

    // [fsi:val s : float32 = 4.14f]
    // [fsi:val f : float = 4.14]
    // [fsi:val d : decimal = 0.7833M]
    // [fsi:val bi : System.Numerics.BigInteger = 9999]

See [Literals (MSDN)](http://msdn.microsoft.com/en-us/library/dd233193.aspx) for complete reference.

<a name="Functions"></a>Functions
---------
The `let` keyword also defines named functions:

    let negate x = x * -1 
    let square x = x * x 
    let print x = printfn "The number is: %d" x

    let squareNegateThenPrint x = 
        print (negate (square x)) 

### Pipe and composition operators

> Pipe operator `|>` is used to chain functions and arguments together. Double-backtick identifiers are handy to create a symbol with spaces in it to improve readability, especially for unit testing:

Author: The note about unit testing seemed like a non sequitur until I decoded it. I think it could read more smoothly. How about “Extra-long identifiers in double-backticks are especially handy for unit test names”.

> Pipe operator `|>` is used to chain functions and arguments together. Double-backtick identifiers are handy to create a symbol with spaces in it to improve readability, especially for unit testing:

Reviewer: ```suggestion
The `|>` pipe operator is used to chain function calls:, passing each result of the preceding function as the argument to the next one:```

I would remove the double tick stuff from this section entirely as it has nothing to do with basics of functions and composing them

>     let ``square, negate, then print`` x = 

Reviewer: needs to go, breaks the flow

>     let ``square, negate, then print`` x = 

Reviewer: ```suggestion
    let squareNegateThenPrint x = 
```
we are trying to convey that its identical to the prev example

        x |> square |> negate |> print
	

> This operator [what operator?] is essential in assisting the F# type checker by providing type information before use:

Author: It‘s ambiguous whether “this” refers to something preceding or something to be unveiled in the following example. Are we talking about the “:” operator? If so, then it should be introduced better. Something like
> The “:” operator can be used to specify the type of a parameter:

> This operator [what operator?] is essential in assisting the F# type checker by providing type information before use:

Reviewer: ```suggestion
Aside from the fact it can read better, `|>` also works well with _type inference_, allowing one to elide explicit type annotations:
```
the example is not a good one (IRL you'd use `Array.sumBy`, there is no useful type propagation, abd it does not show before vs after), but I can't think of a better one rn.

    let sumOfLengths (xs : string []) = 
        xs 
        |> Array.map (fun s -> s.Length)
        |> Array.sum

> Composition operator `>>` is used to compose functions in a way that's more more concise than shown above. Also, appending a backtick to a symbol is a conventional way to mark the new symbol as a variant of a previously defined symbol. It's like appending `1` but cooler:

Author: How about:
> The composition operator `>>` is a crisper way to express a pipeline:
Also, the example muddies the waters by not being an exact analogue to the earlier piping example that uses double-backtick. I suggest introducing double-backtick somewhere else.

Non-functional people will not have seen single quote used this way. They can imagine all sorts of meanings. Saying something explicit about it makes a smoother experience for them. Maybe say it after, instead of before?

> Composition operator `>>` is used to compose functions in a way that's more more concise than shown above. Also, appending a backtick to a symbol is a conventional way to mark the new symbol as a variant of a previously defined symbol. It's like appending `1` but cooler:

Reviewer: Might make sense to add a 'function names' section saying"
- a `'` suffix in a function name is often used to separate pre vs post values
- the ```` double tick quotes can be used to quote names with embedded spaces. This is often used in test code

Either way, no pint making it harder to convey the role and some key pros and cons of `>>` without talking about naming too

> Composition operator `>>` is used to compose functions in a way that's more more concise than shown above. Also, appending a backtick to a symbol is a conventional way to mark the new symbol as a variant of a previously defined symbol. It's like appending `1` but cooler:

Reviewer: Regarding explaining `>>` the best way is to take an exact copy of something above. As long as the text does not digress into ticks and double backticks, that'll be fine
example can be:
```
let squareNegateThenPrint = square >> negate >> print
``

    let squareNegateThenPrint' = 
        square >> negate >> print
  
### Recursive functions

> The `rec` keyword is used together with the `let` keyword to define a recursive function:

Reviewer: ```suggestion
The `rec` keyword is used together with the `let` keyword to define a recursive function (normally you can't refer to the name of a function from within it):
```

maybe an example showing shadowing might explain why you need it?
https://github.com/fsharp/fslang-suggestions/issues/1299

    let rec fact x =
        if x < 1 then 1
        else x * fact (x - 1)

> 

Reviewer: maybe an example showing shadowing ?
https://github.com/fsharp/fslang-suggestions/issues/1299

> *Mutually recursive* functions (functions that call each other) are indicated by `and` keyword:

Reviewer: ```suggestion
*Mutually recursive* functions (functions that call each other) are grouped by using the `and` keyword (instead of `let`):
```

>     let rec even x =

Reviewer: needs a better example - v distracting to read a nonsenensical function compared to doing a `% 2 = 0` check

        if x = 0 then true 
        else odd (x - 1)

     and odd x =
        if x = 0 then false
        else even (x - 1)

<a name="PatternMatching"></a>Pattern Matching
----------------

> Pattern matching is often facilitated through `match` keyword.

Reviewer: ```suggestion
Pattern matching is via the `match` keyword:
```

    let rec fib n =
        match n with
        | 0 -> 0
        | 1 -> 1
        | _ -> fib (n - 1) + fib (n - 2)

> More sophisticated matching can be done with `when`, to create a filter or guard on a pattern:

Reviewer: ```suggestion
Matching has many powerful features, e.g. the `when` clause adds a filter/guard on a match:
```

    let sign x = 
        match x with
        | 0 -> 0
        | x when x < 0 -> -1
        | x -> 1

> If matching is the first thing a function will do, use the `function` keyword to start the match:

Reviewer 2: I don't think this change adds clarity, as "matching is the first thing a function will do" in all examples here, not just the example with `function`.

Author: Maybe I'‘m missing something.  It seems like an alternate way of doing the same thing. Which is preferred?

Reviewer: Its just an alternate syntax, and some people frown on it as it doesnt' save any chars (just one line) and can arguably be harder to scan.
As noted below, convey that its identical by just doing `sign` as a `function`

>     /// Similar to `fib`; using `function` for pattern matching

Reviewer: replace with `sign` as thats directly above and conveys the point fine.

    let rec fib' = function
        | 0 -> 0
        | 1 -> 1
        | n -> fib' (n - 1) + fib' (n - 2)

Pattern matching can also be done directly on arguments:

>     let fst' (x, _) = x

Reviewer: ```suggestion
    let fst (x, _y) = x
```
keep same name to show you are doing an identical impl

> For more complete reference visit [Pattern Matching (MSDN)](http://msdn.microsoft.com/en-us/library/dd547125.aspx).

Reviewer: move up to where the `hen` cluase is shown

### Lists
A *list* is an immutable collection of elements of the same type.

> 

Reviewer: maybe show:

`@` is an alias for `List.append`, and can join two` list`s:
let list3 = list1 @ list 2
